transfers/amclient.py

```diff
@@ -0,0 +1,487 @@
+        else:
+            raise TypeError("expect bytes or str, not %s" %
+                            type(filename).__name__)
+import pprint
+import re
+import sys
+
+import begin
+import requests
+from six.moves import configparser
+
+
+THIS_DIR = os.path.abspath(os.path.dirname(__file__))
+DEFAULT_LOGFILE = os.path.join(THIS_DIR, 'amclient.log')
+LOGGER = logging.getLogger('amclient')
+CONFIG_FILE = None
+RETRY_COUNT = 5
+DEF_AM_URL = 'http://127.0.0.1'
+DEF_SS_URL = 'http://127.0.0.1:8000'
+DEF_USER_NAME = 'test'
+UUID_PATT = re.compile(
+    '^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$')
+
+
+def get_setting(setting, default=None):
+    """Get ``setting`` from our config file; return ``default`` if not
+    possible.
+    """
+    config = configparser.SafeConfigParser()
+    try:
+        config.read(CONFIG_FILE)
+        return config.get('amclient', setting)
+    except (TypeError, configparser.Error):
+        return default
+
+
+def _call_url_json(url, params, method='GET'):
+    """Helper to GET a URL where the expected response is 200 with JSON.
+    :param str url: URL to call
+    :param dict params: Params to pass to requests.get
+    :returns: Dict of the returned JSON or None
+    """
+    if method.upper() in ('GET', 'POST', 'DELETE', 'PUT', 'HEAD', 'OPTIONS'):
+        method = method.upper()
+    else:
+        method = 'GET'
+    LOGGER.debug('URL: %s; params: %s; method: %s', url, params, method)
+    response = requests.request(method, url=url, params=params)
+    LOGGER.debug('Response: %s', response)
+    if not response.ok:
+        LOGGER.warning('%s Request to %s returned %s %s', method, url,
+                       response.status_code, response.reason)
+        LOGGER.debug('Response: %s', response.text)
+        return None
+    try:
+        return response.json()
+    except ValueError:  # JSON could not be decoded
+        LOGGER.warning('Could not parse JSON from response: %s',
+                       response.text)
+        return None
+
+
+def b64decode_ts_location_browse(result):
+    """Base64-decode the results of a call to SS GET
+    /location/UUID/browse/.
+    """
+    def dec(thing):
+        try:
+            thing = base64.b64decode(thing.encode('utf8'))
+        except UnicodeEncodeError:
+            LOGGER.warning('Failed to UTF8-encode output from GET call to SS'
+                           ' /location/UUID/browse/: %s', result)
+        except (binascii.Error, TypeError):
+            LOGGER.warning('Failed to base64-decode file or directory names in'
+                           ' output from GET call to SS'
+                           ' /location/UUID/browse/: %s', result)
+        return thing
+    try:
+        result['directories'] = [dec(d) for d in result['directories']]
+        result['entries'] = [dec(e) for e in result['entries']]
+        result['properties'] = dict((dec(k), v) for k, v in
+                                    result['properties'].items())
```

Comment:
Dictionary comprehensions are pretty great. This can be `{(dec(k): v for k, v in result['properties'].items()}`